src/peer_connection.cpp

```diff
@@ -4692,14 +4707,31 @@ namespace libtorrent {
 		write_have(new_piece);
 
 		if (replace_piece >= piece_index_t(0))
+			m_superseed_pieces.erase(std::remove(m_superseed_pieces.begin(), m_superseed_pieces.end(), replace_piece), m_superseed_pieces.end());
+		m_superseed_pieces.insert(m_superseed_pieces.begin(), new_piece);
+
+		// Check we haven't exceeded the piece limit. This may happen
+		// if a peer filters sending redundant HAVE messsages
+		if(m_superseed_pieces.size() > m_max_superseed_pieces)
 		{
-			// move the piece we're replacing to the tail
-			if (m_superseed_piece[0] == replace_piece)
-				std::swap(m_superseed_piece[0], m_superseed_piece[1]);
+#ifndef TORRENT_DISABLE_LOGGING
+			if (should_log(peer_log_alert::info))
+				peer_log(peer_log_alert::info, "SUPERSEED", "trimming superseeded pieces from %d to %d"
+					, static_cast<int>(m_superseed_pieces.size())
+					, static_cast<int>(m_max_superseed_pieces));
+#endif
+			m_superseed_pieces.resize(m_max_superseed_pieces);
 		}
 
-		m_superseed_piece[1] = m_superseed_piece[0];
-		m_superseed_piece[0] = new_piece;
+	}
+
+	uint32_t peer_connection::target_superseed_pieces() const
+	{
+		std::shared_ptr<torrent> t = m_torrent.lock();
+		torrent_info const& ti = t->torrent_file();
+		int const ps = int(ti.piece_length());
+		TORRENT_ASSERT(ps > 0);
+		return std::max(m_settings.get_int(settings_pack::superseed_scale_time) * m_statistics.upload_payload_rate() / ps, 2);
```

Comment:
I think it's worth a comment here, explaining that we need at least `2` here, in order to bootstrap and offer *any* pieces before we have started uploading (and the upload rate is 0).

In fact, it could probably be 1

Reply:
I left the minimum at 2 as that's the current behavior, but 1 probably makes more sense to start with, especially now that its able to grow.

Reply:
If we only offer a single piece to a peer, then they'll probably send us a 'not interested' after each piece, causing us to prematurely choke them and sending the to the back of the queue. Offering at least 2 pieces may help to prevent this.